src/Reflection/ParametersAcceptorSelector.php

```diff
@@ -48,7 +49,12 @@ public static function selectSingle(
 			throw new ShouldNotHappenException('Multiple variants - use selectFromArgs() instead.');
 		}
 
-		return $parametersAcceptors[0];
+		$parametersAcceptor = $parametersAcceptors[0];
+		if (TypeUtils::containsConditional($parametersAcceptor->getReturnType())) {
+			return new SingleParametersAcceptor($parametersAcceptor);
+		}
```

Comment:
I couldn't immediately get tests to pass without guarding here. I could whip up a helper class that can convert a given `ParametersAcceptor` into the same type but with the return type adjusted? That way the contract still holds.